### DISCUSSION TITLE
question regarding code in parttree

### DISCUSSION THREAD

**Discussion by dibyendumajumdar** [mjhoptics/ray-optics#106] [General]
Hi @mjhoptics 

I hope I am reading this right - it appears to me that parent_node should be obtained from position i-1?

https://github.com/mjhoptics/ray-optics/blob/79aa3ae6d89efee9a749d5448444c38ec668a5af/src/rayoptics/elem/parttree.py#L88

Regards

**Comment by mjhoptics** [comment]
I believe the code is correct as written, although I'll admit it is rather cryptic. The function in question is local to a method used to sort the tree nodes to "match" the order of the sequential model interfaces. (Matching may not be possible for some double pass systems, I'm thinking). I've attached a pdf of a model that has multiple assemblies, as well as doublets and mirrors.  I copy/pasted the code from sort_tree_using_sequence() and added printout for what the function produces. 

Basically, I'm starting at a leaf node (either interface or gap) and want to get a list of parents from the element level up to the root, including any assemblies found along the way. The root is first in the list. When traversing this list (path2root), we're looking for lists of children, so we iterate over path2root[1:]. Indexing starts at 0, however, so the first time thru i=0 and this makes parent_node=root.

The zip file contains the notebook as a script. Hope this helps. Mike

[Issue106_part_tree.pdf](https://github.com/mjhoptics/ray-optics/files/10099482/Issue106_part_tree.pdf)
[Issue106_part_tree.py.zip](https://github.com/mjhoptics/ray-optics/files/10099502/Issue106_part_tree.py.zip)

**Comment by dibyendumajumdar** [comment]
okay my bad - my Python is not great and I should have consulted the docs to see how enumerate() works.

Thank you for the explanation - I have to say the part tree is the one thing I struggle to follow!

Regards

**Reply by mjhoptics** [reply]
A lot of [PartTree](https://ray-optics.readthedocs.io/en/latest/api/rayoptics.elem.parttree.html#rayoptics.elem.parttree.PartTree) is a wrapper around the [anytree](https://anytree.readthedocs.io/en/latest/index.html) package, which implements the tree structure in `parttree`. The search, sort and filtering capabilities of `anytree` are quite good, but can be complicated to code/specify. I tried to make the api simple and useful; I think the use of tags to filter the data in the tree works out well for the data ray-optics models.
There's definitely a part of the `parttree` interface that I use as a user vs. the parts that handle the database/modeling parts.
Mike
